### DISCUSSION TITLE
Invalid Hook Call when implementing a theme plugin

### DISCUSSION THREAD

**Discussion by jy95** [facebook/docusaurus#8571] [Q&A]
**Context**

I am implementing a theme plugin (https://github.com/jy95/docusaurus-json-schema-plugin) and it requires an async task to be completed before rendering the component

```tsx
import React, { useState, useEffect } from "react"
import { Resolver } from "@stoplight/json-ref-resolver"

import { CreateNodes, Collapsible } from "./components/index"

import type { JSONSchema7 } from "json-schema"
import type { IResolverOpts } from "@stoplight/json-ref-resolver/types"

export type Props = {
  schema: JSONSchema7
  // To customize the ref resolving
  resolverOptions?: IResolverOpts
  [x: string]: any
}

type InnerViewerProperties = {
  // Thanks to @stoplight/json-ref-resolver, $ref are either :
  // 1. resolved
  // 2. unresolved (as circular stuff are not on the roadmap)
  schema: JSONSchema7
}

// Internal
function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  const { schema } = props
  // Title of the schema, for user friendliness
  const title = schema?.title || "Schema"
  return (
    <Collapsible
      summary={<strong>{title}</strong>}
      detailsProps={{
        open: true,
      }}
    >
      <CreateNodes schema={schema} />
    </Collapsible>
  )
}

// Entry point
export default function JSONSchemaViewer(props: Props): JSX.Element {
  const { schema: originalSchema, resolverOptions } = props
  const [error, setError] = useState(undefined as undefined | Error)
  const [resolvedSchema, setResolvedSchema] = useState(
    undefined as undefined | JSONSchema7
  )

  useEffect(() => {
    // Set up resolver
    const resolver = new Resolver(resolverOptions)

    // Time to do the job
    resolver
      .resolve(originalSchema)
      .then((result) => {
        setResolvedSchema(result.result)
      })
      .catch((err) => {
        setError(err)
      })
  }, [])

  if (error !== undefined) {
    return <div>Something bad happens : {error.message}</div>
  } else if (resolvedSchema === undefined) {
    return <div>Loading ....</div>
  } else {
    return <JSONSchemaInnerViewer schema={resolvedSchema} />
  }
}
```

However, as soon useState / useEffect  is invoked, I got the infamous "Invalid Hook Call"
![image (1)](https://user-images.githubusercontent.com/9306961/214122350-a89767b0-eeb8-441c-bf4e-d34f7e708a88.png)
![image](https://user-images.githubusercontent.com/9306961/214122359-06683ca6-76e9-4732-bcec-8285659b5696.png)

**Expected**
![image (2)](https://user-images.githubusercontent.com/9306961/214122261-7ffa677f-4ac0-4b12-a121-dc1559a8dec2.png)
Like https://jy95.github.io/docusaurus-json-schema-plugin/docs/demo-viewer/object/additionalProperties (which doesn't use a useState / useEffect )

**Steps to reproduce**
- git clone https://github.com/jy95/docusaurus-json-schema-plugin.git
- git checkout -b statefulComponents 
- npm install
- npm run build
- cd testsite
- npm install
- npm start
- On web, go to /docs/demo-viewer/object/additionalProperties

**Checklist of stuffs I checked**
- npm ls react (only one version : 17.0.2)
- npm ls react-dom (only one version : 17.0.2)
- putting "export default" at the same line that "function ...." (who knows ...)
- try without / without useState & useEffect to see if workings without these hooks (well, these hooks triggered the issue)
- Double check "rules of Hooks" : https://reactjs.org/warnings/invalid-hook-call-warning.html
- (Un)comment configureWebpack() to see if "monaco-editor-webpack-plugin" could explain that (well, on or off didn't change)

**Use cases**

In a docs page

```mdx
---
description: A doc page with my component.
---

import CodeBlock from '@theme/CodeBlock';
import Schema from "@site/static/schemas/examples/object/additionalProperties.json";
import JSONSchemaViewer from "@theme/JSONSchemaViewer";

# AdditionalProperties (1)

JSON Schema :

<CodeBlock language="json">{JSON.stringify(Schema, null, 2)}</CodeBlock>

Viewer :

<JSONSchemaViewer schema={ Schema  } />
```

Or within a page :
 
```tsx
import React from "react"
import Layout from "@theme/Layout"
// @ts-ignore
import JSONSchemaViewer from "@theme/JSONSchemaViewer"
// @ts-ignore
import JSONSchemaEditor from "@theme/JSONSchemaEditor"

import Schema from "@site/static/schemas/realWorld/boxFileSchema.json"

function BoxFileSchema(): JSX.Element {
  return (
    <Layout
      title={`Box File JSON Schema`}
      description="Description will go into a meta tag in <head />"
    >
      <div style={{ display: "flex", flexWrap: "wrap" }}>
        <div style={{ boxSizing: "border-box", width: "50%" }}>
          <JSONSchemaViewer schema={Schema} />
        </div>
        <div style={{ boxSizing: "border-box", width: "50%" }}>
          {/*<JSONSchemaEditor schema={ Schema } />*/}
        </div>
      </div>
    </Layout>
  )
}

export default BoxFileSchema
``` 

Thanks for the help

**Comment by slorber** [comment]
Most likely:


> You might have more than one copy of React in the same app.

`node_modules/..../.../react` + `testsite/node_modules/react`

If your Docusaurus plugin is a theme it should make React a first-class dependency of it, and ensure that there's no duplicate (npm/yarn using hoisting/workspaces).

Even `node_modules/react` + `testsite/node_modules/react` using exactly the same React version would be a wrong setup so you really need to deduplicate React packages here.


---

@RDIL your plugin starter works for most plugins but is likely to lead to such issues for theme users once they start using React hooks 🤪 

You could add a little theme/component setup in it + a hook to ensure it works.
